src/libtriton/format/pe/pe.cpp

```diff
@@ -0,0 +1,220 @@
+**
+**  This program is under the terms of the BSD License.
+*/
+
+#include <cstdio>
+#include <new>
+
+#include <pe.hpp>
+#include <exceptions.hpp>
+
+
+
+namespace triton {
+  namespace format {
+    namespace pe {
+
+      Pe::Pe(const std::string& path) {
+        this->path      = path;
+        this->raw       = nullptr;
+        this->totalSize = 0;
+
+        this->open();
+        this->parse();
+        this->initMemoryMapping();
+        this->initImportTable();
+        this->initExportTable();
+      }
+
+
+      Pe::~Pe() {
+        delete[] this->raw;
+      }
+
+
+      void Pe::open(void) {
+        FILE* fd = nullptr;
+
+        // Open the file
+        fd = fopen(this->path.c_str(), "rb");
+        if (fd == nullptr)
+          throw triton::exceptions::Pe("Pe::open(): Cannot open the binary file.");
+
+        // Get the binary size
+        fseek(fd, 0, SEEK_END);
+        this->totalSize = ftell(fd);
+        rewind(fd);
+
+        this->raw = new(std::nothrow) triton::uint8[this->totalSize];
+        if(!this->raw)
+          throw triton::exceptions::Pe("Pe::open(): Not enough memory.");
+
+        // Read the file contents
+        if (fread(this->raw, 1, this->totalSize, fd) != this->totalSize)
+          throw triton::exceptions::Pe("Pe::open(): Cannot read the binary file.");
+
+        // Close the file
+        fclose(fd);
+      }
+
+
+      bool Pe::parse(void) {
+        this->header.parse(this->raw, this->totalSize);
+        return true;
+      }
+
+
+      void Pe::initMemoryMapping(void) {
+        for (auto section : header.getSectionHeaders()) {
+          triton::format::MemoryMapping area(this->raw);
+
+          triton::uint32 rawAddr = section.getRawAddress();
+          triton::uint32 rawSize = section.getRawSize();
+          triton::uint32 virtAddr = section.getVirtualAddress();
+          if (this->totalSize < (rawAddr + rawSize)) {
+            std::cerr << "Warning Pe::initMemoryMapping(): Some PE Section Headers of the binary file are corrupted." << std::endl;
+            continue;
+          }
+
+          area.setOffset(rawAddr);
+          area.setSize(rawSize);
+          area.setVirtualAddress(virtAddr);
+
+          this->memoryMapping.push_back(area);
+        }
+      }
+
+
+      triton::uint64 Pe::getOffsetFromAddress(triton::uint64 vaddr) const {
+        for (auto section : this->header.getSectionHeaders()) {
+          if (vaddr >= section.getVirtualAddress() && vaddr < (section.getVirtualAddress() + section.getRawSize())) {
+              return ((vaddr - section.getVirtualAddress()) + section.getRawAddress());
+          }
+        }
+        return 0;
+      }
+
+      void Pe::initExportTable(void) {
+          triton::uint32 exportStart = header.getDataDirectory().getExportTable_rva();
+          triton::uint32 exportSize = header.getDataDirectory().getExportTable_size();
+
+          if (exportStart == 0) {
+              std::memset(&exportTable, '\0', 40);
```

Comment:
This was a mistake, after converting PeExportDirectory to a class it initializes itself with zeros so there is no longer a need to have a memset here.